sky/serve/serve_utils.py

```diff
@@ -299,15 +302,21 @@ def update_service_encoded(service_name: str, version: int, mode: str) -> str:
             'mode': mode,
         })
     if resp.status_code == 404:
```

Comment:
I have a question about this error handling. Since `serve_utils` and the endpoints are both executed on the controller, if `update_service` is old and doesn't have that endpoint, the `update_service_encoded` function wouldn't exist either. In other words, if one exists, both should exist. So this 404 case seems unlikely to be triggered. 
Is my understanding correct, or is there a scenario I'm missing where this could happen?

Reply:
That is a good question! The reason is a little bit complicated.

1. Python will load the code in memory after a script is executed, and it wont automatically reload the code after the content in the file changed. You can try this simple fastapi server by running it, uncomment its `/a` endpoint code, and then curl its /a path; it will still shows a 404 not found error. This is the reason why the `'/controller/update_service'` endpoint might be missing for old services.

2. We will update the skypilot code on the cluster for every `sky launch`. For service controller, that would be every `sky serve up`. And our codegen is executed after the code is updated, so it will read the latest code from the file, thus the `update_service_encoded` function can be found and executed.

```python
from fastapi import FastAPI

app = FastAPI()

# @app.get("/a")
# def read_a():
#     return {"message": "Hello from endpoint A!"}

@app.get("/b")
def read_b():
    return {"message": "Hello from endpoint B!"}

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000)
```

Reply:
I see your point, but I'm not sure that's entirely correct. When we do a 'serve up', wouldn't it also update the controller code? If so, the newly started service should have the new endpoints as well. Am I missing something here? Could you clarify how the controller and service code updates are handled during a 'serve up'?

Reply:
Assuming you launched some services before you add the endpoint, then you add those endpoints, do another `serve up`, update the controller code. The newly started service (with this `serve up`) would indeed have the new endpoints, but that is not the case for the services launched before you made this code change.

Reply:
Is this the correct scenario: A new code release happens, then a new service is launched, triggering a SkyPilot runtime code update. The controller process remains running with old code and endpoints, but both old and new services will use the updated code for their operations because they start new processes each time. Is this mismatch between the persistent controller and the on-demand service processes the root cause of the potential endpoint/function discrepancy?

Reply:
> both old and new services will use the updated code for their operations because they start new processes each time.

For this, are you referring to the service codegen and command runner process? If so then it is the correct understanding 🙂